checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInferenceLogic.java

```diff
@@ -43,104 +82,596 @@ public class MustCallInferenceLogic {
+
+    if (!(rhs instanceof LocalVariableNode)) {
+      return;
+    }
+    Obligation rhsObligation =
+        MustCallConsistencyAnalyzer.getObligationForVar(obligations, (LocalVariableNode) rhs);
+    if (rhsObligation == null) {
+      return;
+    }
+
+    if (lhsElement.getKind() == ElementKind.FIELD) {
+      if (!getEnclosedOwningFields().contains(lhsElement)) {
+        return;
+      }
+
+      if (TreeUtils.isConstructor(enMethodTree)) {
+        addOwningToParamsIfDisposedAtAssignment(obligations, rhsObligation, rhs);
+      } else {
+        if (owningFieldToECM.contains((VariableElement) lhsElement)) {
+          owningFieldToECM.remove((VariableElement) lhsElement);
+        }
+        addOwningToParamsIfDisposedAtAssignment(obligations, rhsObligation, rhs);
+      }
+
+    } else if (lhsElement.getKind() == ElementKind.RESOURCE_VARIABLE && mcca.isMustCallClose(rhs)) {
+      addOwningToParamsIfDisposedAtAssignment(obligations, rhsObligation, rhs);
+    } else if (lhs instanceof LocalVariableNode) {
+      LocalVariableNode lhsVar = (LocalVariableNode) lhs;
+      mcca.updateObligationsForPseudoAssignment(obligations, assignmentNode, lhsVar, rhs);
+    }
+  }
+
+  /**
+   * Adds owning fields to the method's parameters, if the must-call obligation of its alias
+   * referred to by the right-hand side of the assignment is fulfilled during the assignment.
+   *
+   * @param obligations The set of obligations to update.
+   * @param rhsObligation The obligation associated with the right-hand side of the assignment.
+   * @param rhs The right-hand side of the assignment.
+   */
+  private void addOwningToParamsIfDisposedAtAssignment(
+      Set<Obligation> obligations, Obligation rhsObligation, Node rhs) {
+    Set<ResourceAlias> rhsAliases = rhsObligation.resourceAliases;
+    for (ResourceAlias rl : rhsAliases) {
+      Element rhdElt = rl.reference.getElement();
+      List<? extends VariableTree> params = enMethodTree.getParameters();
+      for (int i = 0; i < params.size(); i++) {
+        VariableElement paramElt = TreeUtils.elementFromDeclaration(params.get(i));
+        if (paramElt.equals(rhdElt)) {
+          addOwningOnParams(i);
+          mcca.removeObligationsContainingVar(obligations, (LocalVariableNode) rhs);
+          break;
+        }
+      }
+    }
+  }
+
+  /**
+   * Adds an {@link EnsuresCalledMethods} annotation to the enclosing method for any owning field
+   * whose must-call obligation is satisfied within the enclosing method.
    */
-  private void checkForMustCallInvocationOnField(MethodInvocationNode mNode) {
+  private void addEnsuresCalledMethods() {
+    Map<String, Set<String>> map = new LinkedHashMap<>();
+    for (VariableElement owningField : owningFieldToECM) {
+      List<String> mustCallValues = typeFactory.getMustCallValue(owningField);
+      assert !mustCallValues.isEmpty() : "Must-call obligation of an owning field is deleted.";
+      // Assume must-call set has one element
```

Comment:
This class makes the assumption that the must call set has only one element in several places. I think this should be listed in the class' documentation (and probably the user manual, too) as a limitation that we hope to remove in the future.

Reply:
Added.